Eliminate Hashtbl.find_all

mjambon opened the pull request:
This adds:
* replacements for `Hashtbl.find_all` and `Hashtbl.add`: `Hashtbl_.get_stack` and `Hashtbl_.push`.
* a local Semgrep rule to catch future uses of `Hashtbl.find_all`.

Unit tests for the commons library are now grouped into a single module exposed by the commons.tests library.

Commit:
Add safe replacements for Hashtbl.find_all

Commit:
Fix signature

Commit:
Remove all uses of List.find_all (since it's not stack-safe in OCaml < 5)

Commit:
Add an internal Semgrep rule to avoid using Hashtbl.find_all.

aryx commented on libs/commons/Hashtbl_.mli:
```diff
@@ -6,3 +6,32 @@ type 'a hashset = ('a, bool) Hashtbl.t
 val hashset_of_list : 'a list -> 'a hashset
 val hashset_to_list : 'a hashset -> 'a list
+
+(* Safe replacement for Hashtbl_.get_stack for OCaml < 5
```
you mean Safe replacement for Hashtbl.find_all

mjambon replied:
This is one of the dangers of global replace (with sed, not semgrep).

aryx commented on libs/commons/Hashtbl_.mli:
```diff
@@ -6,3 +6,32 @@ type 'a hashset = ('a, bool) Hashtbl.t
+
+(* Safe replacement for Hashtbl_.get_stack for OCaml < 5
+
+   In Ocaml < 5, Hashtbl_.get_stack is not stack-safe and causes Semgrep
```
again Hashtbl.find_all here

aryx commented on semgrep.jsonnet:
```diff
@@ -73,6 +73,17 @@ local semgrep_rules = [
+    severity: 'ERROR',
+    message: |||
+      `Hashtbl.find_all` is not stack-safe in OCaml < 5. Use `Hashtbl_.push`
+      instead of `Hashtbl_add` and `Hashtbl_.get_stack` instead of
```
typo, it's Hashtbl.add here

aryx commented on src/engine/Match_env.ml:
```diff
@@ -31,8 +31,8 @@ module PM = Pattern_match
 
-(* !This hash table uses the Hashtbl.find_all property! *)
-type id_to_match_results = (pattern_id, Pattern_match.t) Hashtbl.t
+(* !This hash table uses the Hashtbl_.get_stack property! *)
```
I don't think you need to put those comments anymore; Previously, I was adding such as comment because
the signature of the hashtbl does not indicate whether the hashtbl should be interpreted as a set of values
or one value. Now that the value type contains list ref, it's clear we're using a set of values for each keys.

mjambon replied:
thanks. I was wondering.

aryx commented on libs/commons/Hashtbl_.mli:
```diff
@@ -6,3 +6,32 @@ type 'a hashset = ('a, bool) Hashtbl.t
+     Hashtbl_.get_stack tbl 42 |> List.rev
+*)
+val push : ('k, 'v list ref) Hashtbl.t -> 'k -> 'v -> unit
+val get_stack : ('k, 'v list ref) Hashtbl.t -> 'k -> 'v list
```
maybe add a comment that if the key is not in the hash, we actually return an empty list, instead of Not_found (like in Hashtbl.find_all). Or maybe return Not_found to better match Hashtbl.find_all.

mjambon replied:
btw, `Hashtbl.find_all` doesn't raise `Not_found` either. I had to check.

```ocaml
utop # Hashtbl.find_all (Hashtbl.create 10) 42;;
- : 'a list = []
```
Adding a comment anyway.

aryx reviewed (approved):
Very nice.

aryx commented:
for reference, similar modif in semgrep-pro done by nat: https://github.com/semgrep/semgrep-proprietary/commit/99938448de1d158c1b41c0e17112abc5225324d7

Commit:
Comments

Commit:
Compile and test the graph_code library
It's used (but untested) by semgrep-proprietary and was failing to
build after recent changes.
